### PR TITLE
allow all JSON types

### PR DESCRIPTION
**Pull request by emiltin**

### CODE REVIEW

**Comment by marcgarba** [comment]
Good news to simply allow to use all JSON types for SXL.
The #135 issue should link to this pull request.

Why removing the specialized "integer" datatype ?
It can be interesting to keep for some informations that are integers only (for performance reason to parse).
For some equipments, many informations can be only integers values (no float)....

**Comment by marcgarba** [comment]
To keep the interested specialized type "integer",
in the list of specialized type, the "base64" and "timestamp" would be send as JSON strings,
but the "integer" would be send as a JSON number.

In the first sentence, it would be good to precise that it is concerning the "signals" exchanged in general.

**Comment by emiltin** [comment]
I've updated this PR to more clearly explain what is:
- basic json types
- specialized types
- legacy string-encoded types

The core spec uses these additional types. Should they be mentioned in the "data types" section as well?

message_id
component_id
version
command_code
status_code
alarm_code
string_list
integer_list
boolean_list

These types each have a definition in our JSON core schema.

**Comment by marcgarba** [comment]
At least,
theses ones :
- string_list
- integer_list
- boolean_list

very useful (for some informations exchanged in SXL in general) would be interesting to add them to the 'specialized types chapter'.

On the other hand for SXL, it is not a good idea to add thinks like CodeId ones, to allow freedom on theses (in France we have planned to use something different with a prefix part in them for organizing status/commands...)   

And perhaps not necessary to add here the types used for RSMP Core part (MessageId, version, ...)

**Comment by marcgarba** [comment]
In "legacy types",
shouldn't 'list_as_string',
splitted in theses three types : 
- string_list_as_string
- integer_list_as_string
- boolean_list_as_string

?
(to get corresponding type)

**Comment by emiltin** [comment]
I added the comma-separated types as suggested, however I named them:
- boolean_string_list
- number_string_list
- integer_string_list

I also simplified boolean_as_string and friends to just:
- number_string
- integer_string
- boolean_string

**Comment by marcgarba** [comment]
I find that using the previous “xxxx_as_string” end termination was more meaningful, for all the legacy types encoded as strings.

Typo in legacy chapter: "legecy" => "legacy",
and "using these specialized types" => "using these legacy types" ?

For better lisibility, it would be better to use the same order each time for : number, integer, boolean (unitary and in lists).

In detail, we could have :
- number_as_string
- integer_as_string
- boolean_as_string
- number_list_comma_separated_as_string
- integer_list_comma_separated_as_string
- boolean_list_comma_separated_as_string
- string_list_comma_separated_as_string

**Comment by marcgarba** [comment]
To complete, wouldn't it be interesting to add a new sub-chapter with **“type combinations”** that can be found classically ? :

- array_of_objects (what is is actually usable in RSMP simulators, and used in TLC for "S0033", "S0005", "S0035")
- array_of_strings
- array_of_numbers
- array_of_integers
- array_of_booleans

Even if potentially everything would be possible in JSON for SXL, this may indicate at least some current cases that could be managed with the tools...

**Comment by emiltin** [comment]
> I find that using the previous “xxxx_as_string” end termination was more meaningful, for all the legacy types encoded as strings.

I renamed the types to:

- number_as_string
- integer_as_string
- boolean_as_string.
- string_list_as_string
- number_list_as_string
- integer_list_as_string
- boolean_list_as_string

I think it's a bit wordy, but maybe being more descriptive is good in this case.

Also fixes the typos.


> wouldn't it be interesting to add a new sub-chapter with “type combinations” that can be found classically ? 
Since these are just combinations of plain JSON types, I don't think we need to describe these. 

For example, S0005 uses this structure:

```json
             "s": [{
                     "intersection": "1",
                     "startup": "True"
             },{
                     "intersection": "2",
                     "startup": "False"
             }]
```

This is a bit convoluted since `{"1": true, "2": false }` would have conveyed the same data. But regarding the types, it's just a json array of json objects. Object keys are integer_as_string, while the values are boolean_as_string.

BTW, note that you cannot use integers as keys in json objects, only strings are allowed.

**Comment by marcgarba** [comment]
Ok, fine.

Just seen this type for number_as_string :  "Follows **s**the ECMA", an extra 's'. 

And not sure here : "Signal Exchange Lists (SXLs) specifiy" => specify ?

**Comment by emiltin** [comment]
We need to go through the core spec and make sure the updates type names are used consistently.

**Comment by emiltin** [comment]
Note: we will keep using the legacy types in the core spec, not to break anything. But plain json types can now we used in new SXLs or SXL messages.

**Comment by emiltin** [comment]
> Ok, fine.
> 
> Just seen this type for number_as_string : "Follows **s**the ECMA", an extra 's'.
> 
> And not sure here : "Signal Exchange Lists (SXLs) specifiy" => specify ?

Fixed the two typos.

**Comment by emiltin** [comment]
We do not plan to use native json types throughout the core spec in 3.3.0, right? Simply allow them in SXLs?
I wish we just used plain json types thoughout, but guess that would be a breaking change.

**Comment by emiltin** [comment]
@otterdahl I went through basic_structure.rts and change relevant tables etc. to use our new data type definations. Would you mind taking a look?

**Comment by marcgarba** [comment]
Since the RSMP Core doesn't change, for the "sOc" element in StatusSubscribe message, we have a 'boolean' type here, and not a 'boolean_as_string'. I guess it was introduced more recently that "uRt" which is effectively a 'number_as_string' type.

**Comment by emiltin** [comment]
> Since the RSMP Core doesn't change, for the "sOc" element in StatusSubscribe message, we have a 'boolean' type here, and not a 'boolean_as_string'. I guess it was introduced more recently that "uRt" which is effectively a 'number_as_string' type.

You're right, fixed.

**Comment by emiltin** [comment]
@otterdahl care to review this? then I think we're ready to merge to 3.3.0 branch.

**Comment by emiltin** [comment]
It would be good to verify that the RTS tables render correctly.

**Comment by otterdahl** [comment]
> It would be good to verify that the RTS tables render correctly.

I've finished reviewing and committed some fixes.

Should we update the SXL to use the legacy data types?

**Review by otterdahl — APPROVED** [review]
